### DISCUSSION TITLE
Have one Fairtravelers codebase with multiple communities

### DISCUSSION THREAD

**Discussion by lennartvanlaake** [OpenHospitalityNetwork/fedi-trustroots#76] [Ideas]
To continue discussion about how to separate all the "branding"/unique code from the applicative/common code, so multiple communities can easily use and pull from a single codebase. Started from #16.

My ansers to @weex comments there:

> > * everything a new community might want to change in terms of text or branding should not be in the (main) repo
> > * a mechanism should exist to inject new static text and branding in a way that allows pulling updates from the (main) repo to one of the community-specific instances
> 
> Feel free to correct me but I think what you mean by these is that the configuration should not be overwritten by a `git pull`. Often this is done by storing such data in db or in a (possibly generated) config file that is .gitignored.

Those are both valid strategies but neither of them are being executed at the moment. And both would require very substential number of lines of codes. So my concern is that if we close #16 we forget to do all that work.

A different solution might just be to not have any frontend in the Fairtravelers codebase.

 Since all branding and text lives in the frontend anyway, each community would just be able to have their own frontend while we provide a backend that provides all the core functionality of a hospex platform and all the functionality to federate. I

I think this solution would make it considerably easier for smaller platforms like Welcometomygraden to hop onboards, since they would get to keep 100% of their look and feel while not having to worry about maintaince of their backend code anymore. It would require a major rewrite of the Trustroots codebase though, but I think that needs to happen anyway for us to do any of the things we want.

> 
> > * all features (such as circles) that are not relevant to all communities should be toggle-able
> 
> This would be good to add in a design philosophy doc. Optionality can be accomplished in many ways like checkboxes, plugins, config flags which are to me all great because the software can be made to work for many different communities.
>

I think we then need to create/think about issues outside of #16 to introduce that in the code somehow :)

**Comment by weex** [comment]
> Those are both valid strategies but neither of them are being executed at the moment.

The demo server is using https://github.com/OpenHospitalityNetwork/fedi-trustroots/blob/master/config/env/local.sample.js which was saved as local.js and allows for the app.title and app.description to be updated and probably many other things. To test, I just did a `git pull` on the demo server and those survived. That's not to say we're 100% out of the woods, but it's good to be specific and list things that should be more generic on the demo server in an issue.

> A different solution might just be to not have any frontend in the Fairtravelers codebase.

While an option, we'd definitely want the codebase to be usable out of the box. The api that's necessary for react/angular front-ends lends itself well to swapping out front-ends but the issue of hard-coding just gets moved into another repo. Again, I don't know the scope of this problem but look forward to tackling in specific issues (the smaller the better).

**Reply by lennartvanlaake** [reply]
> > Those are both valid strategies but neither of them are being executed at the moment.
> 
> The demo server is using https://github.com/OpenHospitalityNetwork/fedi-trustroots/blob/master/config/env/local.sample.js which was saved as local.js and allows for the app.title and app.description to be updated and probably many other things. To test, I just did a `git pull` on the demo server and those survived. That's not to say we're 100% out of the woods, but it's good to be specific and list things that should be more generic on the demo server in an issue.
> 

I would imagine that if you are a community that would want a different layout or completly different text that the merge would become significantly more problematic, and I think most communities would like that. So while removing the literal word "Trustroots" is a nice start, I don't think that would be enough to make our product something that all communities would like.

> > A different solution might just be to not have any frontend in the Fairtravelers codebase.
> 
> While an option, we'd definitely want the codebase to be usable out of the box. The api that's necessary for react/angular front-ends lends itself well to swapping out front-ends but the issue of hard-coding just gets moved into another repo. Again, I don't know the scope of this problem but look forward to tackling in specific issues (the smaller the better).

I would imagine a new community having a backend that they can just use without any adjustments (i.e. just pull a docker image or something) and a frontend that they could fork from somewhere and fill in to their desires. Since the backend should be completely community-neutral, that would require 0 extra work for them and could just be upgraded directly from a package of some sort, unless the community would like a specific feature added, at which point they would open a PR to fairtravelers. 

If they would like to fork the "example" frontend repo and basically keep everything as is, they can pull from the repo and fix any potential merge conflicts when they want pull. If they want to fork it and completely rework it or just use something completely unrelated, that would also be fine, but then the upkeep would their problem :)

Probably some sort of docker-compose can be used to get the "works out of the box" experience".

**Reply by weex** [reply]
> make our product something that all communities would like.

I worry about chasing the impossible. There's that saying about pleasing all of the people... at the same time we should be 100% open and inviting to any community who wants to use the software but can't, to help them create issues and get it going. If you want to stand up an instance but don't want to use the default front-end or some piece of the server code or want literally any other change, you'd make an issue, maybe solve it or someone else would and so on.

**Reply by weex** [reply]
I'd add a helpful exercise that I haven't gone through with this project is to try and imagine myself as a community member or future organizer, taking a look at the entrypoints to the project to understand if there's a path for me. The right features are certainly important in this respect, but can easily be hidden behind pages that don't speak to them.

**Reply by weex** [reply]
> Probably some sort of docker-compose can be used to get the "works out of the box" experience".

I've fallen in love with docker-compose so definitely support this as an easy install option. I've seen setups that do everything including a web server with letsencrypt to just running the app so there's a lot of flexibility but it can also be a lot to support.

**Comment by mariha** [comment]
It has been already proposed to [adapt WTMG frontend to our backend](https://github.com/OpenHospitalityNetwork/OpenHospitalityNetwork.github.io/issues/15), what at the time seemed to me mutually beneficial, as WTMG planned to rewrite their backend entirely. They've decided to go their own way as for now, but it could be an interesting option for some of the communities in the future. Also if anyone was willing to you can ask again, I feel like I've already tried to push the idea too much. 

As for now, I'd focus primarily on bicycle touring community and their needs. Then with some federated features in place, next to Trustroots we could try to onboard some of Facebook groups (Host a sister for example).

But let's also discuss it with others on Monday.
